mysql-test/suite/rocksdb/t/ttl_primary_read_filtering.test

```diff
@@ -0,0 +1,341 @@
+--source include/have_rocksdb.inc
+
+--disable_warnings
+DROP TABLE IF EXISTS t1;
+--enable_warnings
+
+# The purpose of read filtering for tables with TTL is to ensure that during a
+# transaction a key which has expired already but not removed by compaction
+# yet, is not returned to the user.
+#
+# Without this the user might be hit with problems such as disappearing rows
+# within a transaction, etc, because the compaction filter ignores snapshots
+# when filtering keys.
+
+# Basic read filtering test
+CREATE TABLE t1 (
+  a int PRIMARY KEY
+) ENGINE=rocksdb
+COMMENT='ttl_duration=1;';
+
+INSERT INTO t1 values (1);
+INSERT INTO t1 values (2);
+set global rocksdb_force_flush_memtable_now=1;
+
+--real_sleep 1
+--sorted_result
+SELECT * FROM t1;
+
+select variable_value into @c from information_schema.global_status where variable_name='rocksdb_rows_expired';
+set global rocksdb_compact_cf='default';
+select variable_value-@c from information_schema.global_status where variable_name='rocksdb_rows_expired';
+
+DROP TABLE t1;
+
+# Test that some rows are hidden but others aren't...
+CREATE TABLE t1 (
+  a int PRIMARY KEY,
+  b BIGINT UNSIGNED NOT NULL
+) ENGINE=rocksdb
+COMMENT='ttl_duration=5;';
+
+INSERT INTO t1 values (1, UNIX_TIMESTAMP());
+--real_sleep 6
+INSERT INTO t1 values (2, UNIX_TIMESTAMP());
+INSERT INTO t1 values (3, UNIX_TIMESTAMP());
+
+set global rocksdb_force_flush_memtable_now=1;
+
+# 1 should be hidden even though compaction hasn't run.
+--sorted_result
+SELECT a FROM t1;
+
+set global rocksdb_compact_cf='default';
+
+# none should be hidden yet, compaction runs but records arent expired
+--sorted_result
+SELECT a FROM t1;
+
+# all should be hidden now, even though compaction hasnt run again
+--real_sleep 6
+--sorted_result
+SELECT a FROM t1;
+
+DROP TABLE t1;
+
+# Test the filtering code explicitly.
+CREATE TABLE t1 (
+  a int PRIMARY KEY
+) ENGINE=rocksdb
+COMMENT='ttl_duration=1;';
+
+INSERT INTO t1 values (1);
+INSERT INTO t1 values (3);
+INSERT INTO t1 values (5);
+INSERT INTO t1 values (7);
+--real_sleep 2
```

Comment:
yes, I believe so.  If the ttl_duration is 1 second, and we sleep for 2 seconds, there's no possible way it won't be filtered during on the read path.  The snapshot opened by the select will be guaranteed to be 2 seconds after the last insert.  Even if real_sleep 2 doesn't sleep for a full 2 seconds for any reason, it at least sleeps for 1000ms.